### ISSUE TITLE
Open() and close() not called equally often (combine related)

### ISSUE THREAD

**Issue by andrehacker** [stratosphere/stratosphere#314]
Problem: The system sometimes calls open() and close() for reduce different amount of times. Given DOP1 and our wordcount example, open() is called 3-times and close() only twice. Both should be called twice, first for combine and then for reduce. See logs below.

First analysis: The first open() is called by DataSourceTask for combine (probably because InputFormat->Map->Combine are chained all together and thus executed from DataSourceTask). The second and third calls are coming from RegularPactTask (which executes the reduce). See the stack traces below.

How to reproduce:
Add open() and close() method to Reduce in WordCount example, add System.out.println(...); in the first line, set DOP to 1 and run using LocalExecutor.

**Log with Combine:**
Map: open
**Reduce: open** (this is probably the call for combine)
Map: map
Map: map
Map: map
Reduce: combine
Reduce: reduce
Reduce: combine
Reduce: reduce
Reduce: combine
Reduce: reduce
Reduce: combine
Reduce: reduce
Reduce: combine
Reduce: reduce
Reduce: combine
Reduce: reduce
Reduce: combine
Reduce: reduce
Map: close
**Reduce: close**
**Reduce: open**
**Reduce: open** (this is one call too much here)
Reduce: reduce
Reduce: reduce
Reduce: reduce
Reduce: reduce
Reduce: reduce
Reduce: reduce
Reduce: reduce
**Reduce: close**

**Log without combine**
Map: open
Map: map
Map: map
Map: map
Map: close
**Reduce: open**
Reduce: reduce
Reduce: reduce
Reduce: reduce
Reduce: reduce
Reduce: reduce
Reduce: reduce
Reduce: reduce
**Reduce: close**
#### Stack Traces

**First open()**

```
WordCount$CountWords.open(Configuration) line: 116  
RegularPactTask<S,OT>.openUserCode(Stub, Configuration) line: 1212  
SynchronousChainedCombineDriver<T>.openTask() line: 104 
RegularPactTask<S,OT>.openChainedTasks(List<ChainedDriver<?,?>>, AbstractInvokable) line: 1253  
DataSourceTask<OT>.invoke() line: 119   
RuntimeEnvironment.run() line: 344  
Thread.run() line: 744  
```

**Second open()**

```
WordCount$CountWords.open(Configuration) line: 116  
RegularPactTask<S,OT>.initInputLocalStrategy(int) line: 730 
RegularPactTask<S,OT>.initLocalStrategies(int) line: 578    
RegularPactTask<S,OT>.invoke() line: 275    
RuntimeEnvironment.run() line: 344  
Thread.run() line: 744
```

**Third open()**

```
WordCount$CountWords.open(Configuration) line: 116  
RegularPactTask<S,OT>.run() line: 361   
RegularPactTask<S,OT>.invoke() line: 291    
RuntimeEnvironment.run() line: 344  
Thread.run() line: 744  
```

**Comment by StephanEwen** [comment]
I think I know what is happening there:

The combiner is instantiated twice. Once as the combiner before the network data transfer. Second as the combiner on the Reduce side that preaggregates data in case the sort spills to disk. If the sort never spills, then the second combiner never gets called.

I suspect that it is the combiner inside the sorter that does not get closed.

**Comment by StephanEwen** [comment]
Fixed in cf9fad89c2cdc3532237308257ecb45c2dbf7a65
